### ISSUE TITLE
ui flickering after kernel shutdown then restart/reconnect 

### ISSUE THREAD

**Issue by aweng98** [DataBiosphere/leonardo#197]
Notebook UI keeps flickering briefly on kernel restart after kernel shutdown completely. When I reloaded/refreshed the page, flickering then stoped.

**Comment by aweng98** [comment]
![screen shot 2018-02-15 at 11 24 02 am](https://user-images.githubusercontent.com/35533885/36267846-ed25f4dc-1242-11e8-9c52-35426dbc00b0.png)

**Comment by ansingh7115** [comment]
This issue seems not to happen on Safari, interestingly

**Comment by rtitle** [comment]
I am not sure of the solution yet, but I think I narrowed down the issue a bit: it's something in the akka-http proxy. My reasoning:

1. I tested with Jupyter directory (via `run-jupyter-local.sh`) -- the flicker does not happen
2. I tested with Jupyter + Apache -- the flicker does not happen
3. I tested with Jupyter + Apache + akka-http -- the flicker happens

I also looked at some debug logs but it's not too informative. On the Jupyter server side I see the websocket connection continually initialize and close, and on the browser side I see websocket connections open then die with `Opcode: -1` frames. Probably the browser is killing/retrying the connection when it gets this opcode.

The question is, what is happening in the akka-http layer to cause this? Perhaps it is an issue with Jupyter swapping out the websocket too fast -- akka-http thinks the _old_ websocket flow is still valid, when in fact it should be killed and reinitialized from the client.

I'm toying with the idea of using an akka-http [KillSwitch](https://doc.akka.io/docs/akka/2.5/stream/stream-dynamic.html) in the websocket flow. The idea being that we'd cache these kill switches by kernel ID, and invoke the correct one if we get a DELETE request for a given kernel. Need to figure out how to actually do this in the akka-stream API.
